crates/test-helpers/macros/build.rs

```diff
@@ -183,6 +184,106 @@ fn main() {
+                let dst = out_dir.join(file);
+                fs::write(dst, contents).unwrap();
+            }
+
+            let path = PathBuf::from(env::var_os("WASI_SDK_PATH").expect(
+                "point the `WASI_SDK_PATH` environment variable to the path of your wasi-sdk",
+            ));
+            let mut cmd = Command::new(path.join("bin/clang"));
+            let out_wasm = out_dir.join("c_utf16.wasm");
+            cmd.arg("--sysroot").arg(path.join("share/wasi-sysroot"));
+            cmd.arg(c_impl)
+                .arg(out_dir.join("imports.c"))
+                .arg(out_dir.join("imports_component_type.o"))
+                .arg(out_dir.join("exports.c"))
+                .arg(out_dir.join("exports_component_type.o"))
+                .arg("-I")
+                .arg(&out_dir)
+                .arg("-Wall")
+                .arg("-Wextra")
+                .arg("-Werror")
+                .arg("-Wno-unused-parameter")
+                .arg("-mexec-model=reactor")
+                .arg("-g")
+                .arg("-o")
+                .arg(&out_wasm);
+            println!("{:?}", cmd);
+            let output = match cmd.output() {
+                Ok(output) => output,
+                Err(e) => panic!("failed to spawn compiler: {}", e),
+            };
+
+            if !output.status.success() {
+                println!("status: {}", output.status);
+                println!("stdout: ------------------------------------------");
+                println!("{}", String::from_utf8_lossy(&output.stdout));
+                println!("stderr: ------------------------------------------");
+                println!("{}", String::from_utf8_lossy(&output.stderr));
+                panic!("failed to compile");
+            }
+
+            let stem = test_dir.file_stem().unwrap().to_str().unwrap().to_string();
+
+            // Translate the canonical ABI module into a component.
+            let module = fs::read(&out_wasm).expect("failed to read wasm file");
+            let mut encoder = ComponentEncoder::default();
+            encoder.encoding = StringEncoding::UTF16;
+            let component = encoder
+                .module(module.as_slice())
+                .expect("pull custom sections from module")
+                .validate(true)
+                .adapter_file(&wasi_adapter)
+                .expect("adapter failed to get loaded")
+                .encode()
+                .expect(&format!(
+                    "module {:?} can be translated to a component",
+                    out_wasm
+                ));
+            let component_path = out_dir.join("c_utf16.component.wasm");
+            fs::write(&component_path, component).expect("write component to disk");
+
+            wasms.push((
+                "c_utf16",
+                stem,
+                out_wasm.to_str().unwrap().to_string(),
+                component_path.to_str().unwrap().to_string(),
+            ));
+        }
+    }
```

Comment:
I think this `if` is basically copy/pasted from above, so could this be refactored instead to a separate function to reduce the duplication here?